batch/sql/finalize-job-groups.sql

```diff
@@ -0,0 +1,514 @@
+DELIMITER $$
+
+DROP TRIGGER IF EXISTS attempts_after_update $$
```

Comment:
Does the entire SQL file happen in a transaction? I'm just a bit apprehensive about the trigger changes coming before the schema changes.

Reply:
That was my intention, but I probably need to add a `START TRANSACTION`. In the past, I thought the `START TRANSACTION` in the script wasn't honored. I'm a bit worried about the database being able to handle that many updates to the database in one transaction. This is how we call the migration script `mysql --defaults-extra-file={mysql_cnf_file} <{script}`

Reply:
The trigger changes should be fine. The job_group_id field is already in there.

Reply:
I'm more worried about the migration failing halfway through and then the SQL commands aren't idempotent.

Reply:
Apparently transactions [cannot undo DDL operations](https://dev.mysql.com/doc/refman/8.0/en/cannot-roll-back.html) so we cannot rollback the ALTER TABLEs anyway.

OK, looking carefully again, I see that you're just making things not null and adding some columns that are only used in Python. In general, I think we should organize DDLs (like ALTER TABLE) first since they modify what the tables mean, but, yeah, in this case, it seems OK.

I don't have any reason to believe START TRANSACTION is not honored by mysql. I have used it several times at the mysql command line to test things.

Tell me more about your migration concerns? What could go wrong and what bad states do we need to anticipate?